### PR TITLE
[#163776583] Allow new options, including 'after', 'fetchLinks', 'page', and 'pageSize'

### PR DESCRIPTION
**Pull request by newtrat**
[PT-#163776583](https://www.pivotaltracker.com/story/show/163776583)

## Description
`Prismic.everything_search_form/1`, `Prismic.all/1`, and `Prismic.documents/2` now support passing arbitrary parameters to Prismic.

In particular, we can now use the pagination params `after`, `page`, and `pageSize`, as well as `fetchLinks`.

## Motivation and Context

Clients can use [paginated results](https://prismic.io/docs/rest-api/query-the-api/pagination-for-results). They can also get more use out of [linked documents](https://prismic.io/docs/rest-api/query-the-api/fetch-linked-document-fields), since they can get some fields from linked documents without needing to make an extra request.

## Testing
- Use this branch locally
- Edit `dev.exs` to use your repo URL: `config :prismic, repo_url: "<YOUR URL HERE>"`
- Run `iex -S mix`
- Make some requests that use the pagination parameters:
  - `pageSize` is the number of results that will be returned
  - `page` is the page number that you're on (1 for the first page)
  - `after` is the ID of a document; Prismic will only start counting results after that document
  - So one example query is `Prismic.all(%{after: "<SOME ID HERE>", page: "1", pageSize: "5"})`
- To see `fetchLinks` in action, you'll need to get documents that link to other Prismic documents.
  - `fetchLinks` uses the same syntax as it does in [Prismic's API](https://prismic.io/docs/rest-api/query-the-api/fetch-linked-document-fields)
  - For TRR users, one example is `Prismic.documents(%{type: "taxon-size-group"}, %{fetchLinks: "measurement.display_name,measurement.description"})`

## Type of Changes

- [ ] Bug fix _(a non-breaking change which fixes an issue)_
- [x] New feature _(a non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that changes existing functionality)_
- [ ] Bump dependencies
- [ ] Automated Testing / Missing tests
- [ ] Documentation

## Checklist:

- [x] Yes, I have added tests to cover my changes.
- [ ] No, We don't have a test suite in this project yet.
- [ ] My change requires a change to the documentation.
- [ ] My change requires release strategy _(uncomment the Release strategy below)_

cc/ @TheRealReal/buyer-purchase

### CODE REVIEW

File: lib/search_form.ex [hunk 1]
```diff
@@ -13,19 +14,28 @@ defmodule Prismic.SearchForm do
...
       |> build_default_data()
       |> Map.merge(data)
 
-    struct(__MODULE__, api: api, form: form, data: default_data)
+    ref =
+      cond do
+        preview_token = data[:preview_token] -> %Ref{ref: preview_token}
+        ref_label = data[:ref] -> ref_label
+        true -> "Master"
+      end
```
Comment (charlesoft) [inline lib/search_form.ex]: WDYT of:

```elixir
defp find_ref(%{preview_token: nil, ref: nil}), do: "Master"
defp find_ref(%{preview_token: nil, ref: ref}), do: ref
defp find_ref(%{preview_token: preview_token, ref: _ref}), do: preview_token
```

File: test/parser_test.exs [hunk 2]
```diff
@@ -43,12 +46,67 @@ defmodule Prismic.ParserTest do
...
+    type: "measurement",
+    uid: nil
+  }
+
+  describe "parse_document_link/1" do
+    test "translates Prismic v1 response format into DocumentLink struct" do
+      assert Parser.parse_document_link(@prismic_document_link) == @parsed_document_link
+    end
+
+    test "includes fragments in DocumentLink when Prismic responseincludes linked data" do
```
Comment (charlesoft) [inline test/parser_test.exs]: `response includes` 💅 .

File: lib/search_form.ex (hunk 1 above)
Comment (newtrat) [reply lib/search_form.ex]: Great suggestion, thanks

File: test/parser_test.exs (hunk 2 above)
Comment (newtrat) [reply test/parser_test.exs]: I disagree here, or at least I think it's a matter of opinion. I'm used to starting tests with a present-tense verb, thanks to experience with RSpec where tests look like `it "does something"`. It doesn't read quite as nicely with ExUnit but I think it's still ok.

Comment (desmondgee) [reply test/parser_test.exs]: I think he means this part "responseincludes"

Comment (newtrat) [reply test/parser_test.exs]: Oh man! Sorry! 🤦‍♂️Thanks

File: lib/parser.ex [hunk 3]
```diff
@@ -104,7 +104,7 @@ defmodule Prismic.Parser do
         %{value: %{document: %{slug: slug, type: doc_type, data: doc_data} = doc} = value} = _json
       ) do
     fragments =
-      if raw_frags = doc_data[doc_type] do
+      if raw_frags = doc_data[String.to_atom(doc_type)] do
```
Comment (desmondgee) [inline lib/parser.ex]: Was this a bugged?

Comment (newtrat) [reply lib/parser.ex]: Yes. When we translate JSON back to an elixir struct, we convert string keys to atoms, but we don't convert values. So `doc_type` (a value) is a string, but the keys in `doc_data` are all atoms.

File: test/parser_test.exs (hunk 2 above)
Comment (charlesoft) [reply test/parser_test.exs]: Np 😆

**Pull request merged**